tests/kafkatest/tests/replication_test.py

```diff
@@ -15,45 +15,86 @@
 
 from ducktape.tests.test import Test
 from ducktape.utils.util import wait_until
+from ducktape.mark import matrix
+from ducktape.mark import parametrize
 
 from kafkatest.services.zookeeper import ZookeeperService
 from kafkatest.services.kafka import KafkaService
 from kafkatest.services.verifiable_producer import VerifiableProducer
 from kafkatest.services.console_consumer import ConsoleConsumer
 
-import signal
 import time
 
 
+"""
+3 brokers, 3 topics, 3 partitions, 3 replicas on each topic
+min.insync.replicas == 2
+ack -1
+failures: [leader, follower, controller] X [clean, hard, soft]
+One test cases with with ack 1
+One test case with compression toggled on
+
+"""
+
+# Failure types
+CLEAN_BOUNCE = "clean_bounce"
+HARD_BOUNCE = "hard_bounce"
+SOFT_BOUNCE = "soft_bounce"
+
+CLEAN_KILL = "clean_kill"
+HARD_KILL = "hard_kill"
+SOFT_KILL = "soft_kill"
+
+# node types
+LEADER = "leader"
+CONTROLLER = "controller"
+
+
 class ReplicationTest(Test):
     """Replication tests.
     These tests verify that replication provides simple durability guarantees by checking that data acked by
     brokers is still available for consumption in the face of various failure scenarios."""
 
     def __init__(self, test_context):
-        """:type test_context: ducktape.tests.test.TestContext"""
         super(ReplicationTest, self).__init__(test_context=test_context)
 
-        self.topic = "test_topic"
+        self.replication_factor = 3
         self.zk = ZookeeperService(test_context, num_nodes=1)
-        self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, topics={self.topic: {
-                                                                    "partitions": 3,
-                                                                    "replication-factor": 3,
-                                                                    "min.insync.replicas": 2}
-                                                                })
-        self.producer_throughput = 10000
-        self.num_producers = 1
-        self.num_consumers = 1
+        self.topics = {"topic1": {
+                            "partitions": 3,
+                            "replication-factor": self.replication_factor,
+                            "min.insync.replicas": 2},
+                       "topic2": {
+                            "partitions": 3,
+                            "replication-factor": self.replication_factor,
+                            "min.insync.replicas": 2},
+                       "topic3": {
+                            "partitions": 3,
+                            "replication-factor": self.replication_factor,
+                            "min.insync.replicas": 2}
+                    }
+        self.topic_names = ["topic1", "topic2", "topic3"]
```

Comment:
I'm not worried about a little more complexity, but it isn't clear just from reading the code what's going on so it probably at least needs some comment. Also, are we actually *guaranteed* one of them is going to be a follower some of the time, or is it possible we get unlucky with leadership and always manage to kill the leaders of all topics?

These tests require pretty deep understanding of how Kafka works and sometimes rely on assumptions that could change over time (e.g., how partition assignment works, guarantees about distribution of partition leadership, etc). We have some good info lower down in this file about the details about how we do the verification. With tests like these it'd also be good to have methodology info in the class docstring.